### DISCUSSION TITLE
Dynamic IL - Is this out of date?

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/runtime#100621] [General]
Years ago (2009) [I coded this](https://github.com/dotnet/runtime/discussions/100487#discussioncomment-8976548) and it still works, but I wanted to ask is this "current" would something like this be coded the same way in .Net 8? I'm trying to get an idea of how dynamic IL has changed (if at all) over the past fifteen years.

**Comment by huoyaoyuan** [comment]
The IL itself doesn't change. However, there are several things to mention:

- Certain IL streams are not fully valid, but tolerated by old versions. There's no guarantee that they will continue to work. Generally, be explicit about type casting.
- Dynamic methods are not compatible with NativeAOT. There was an experiment to support interpreter with NativeAOT, but far from production.
- Your class manipulation code appears to have GC hole as the objects are not properly pinned. `Unsafe.ByteOffset` is more reliable.

**Reply by The-Futurist** [reply]
Thank you Huo, but the offset of a struct _member_ is static, doesn't vary (certainly not at runtime) and this approach requires a struct _instance_, it would be nicer if there was a way to do this statically, compile time, which is where the `offsetof` operator idea for C# came from (I know that my own solution is not compile time, but at least it's _type_ based). 

Also how is one meant to use this to get the offset of some field _within_ a struct, when the args are **both** of type `T`?

**Reply by CyrusNajmabadi** [reply]
> I [tried finding information](https://github.com/dotnet/csharplang/discussions/6105) about he way to do this but didn't get clarity until now!

I'm glad the suggestion to come to the runtime for this information helped out! :)

**Reply by CyrusNajmabadi** [reply]
>  it would be nicer if there was a way to do this statically, compile time,

There's no way to do this as compile-time, because this value is not a compile-time constant.  It can and will vary based on the decisions a particular runtime may make.

**Reply by CyrusNajmabadi** [reply]
> Also how is one meant to use this to get the offset of some field within a struct, when the args are both of type T?

Can you give an example of a such a struct so i'm on the same page as you as to what you're asking about?  Thanks! :)

**Reply by The-Futurist** [reply]
OK I get what you say about "compile time". So there might be some merit in providing a way to enforce a fixed layout/offset, like today we have an attribute that forces the order of managed fields to be as coded, lexical order, so perhaps a forced layout might have some potential. Like "I don't care what the runtime would like to do here, we need this to be platform independent, I want this field at runtime to be at offset X no matter what the runtime scenario might be".

Here's a struct that this issue stems from:

```cs
namespace Steadsoft.Memory.Heap
{
    [StructLayout(LayoutKind.Sequential)]
    internal unsafe ref struct HeapBlock   // we must only ever access this through a pointer aligned to 16 bytes
    {
        public BlockState State;
        public int ThisSize;
        public int PrevSize;
        public Alignment Alignment;
        public long NextFreeBlockOfx;
        public long PrevFreeBlockOfx;
        public static int UserDataOfx = 16;
        public static int BlockSize = sizeof(HeapBlock);
        /// <summary>
        /// Returns the address that the caller will get when 
        /// this block is allocated to them.
        /// </summary>
        internal unsafe byte* UserDataPtr
        {
            get { fixed (long* ptr = &NextFreeBlockOfx) { return (byte*)ptr; } }
        }
        /// <summary>
        /// Give an arbitrary address determines it alignment.
        /// </summary>
        /// <returns>The maximum alignment we can use for this address.</returns>
        internal Alignment UserAlignment
        {
            get
            { 
                var position = BitOperations.TrailingZeroCount((long)(UserDataPtr));
                return position > 14 ? (Alignment)(1 << 14) : (Alignment)(1 << position);
            }
        }
    }
}
```
All of the fields beneath `Alignment` are used as storage space until the caller frees the memory, at which time they become use as to form a linked list. The size of the storage of the first four fields plays a critical role in the internal calculations used within the heap itself, granted this is an unusual scenario but it is real, and I know you guys have used C# to write an operating system too, so this isn't out of this world.

**Reply by CyrusNajmabadi** [reply]
>  I want this field at runtime to be at offset X no matter what the runtime scenario might be.

Can you clarify what is insufficient about https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.fieldoffsetattribute?view=net-8.0 ?

Thanks! :)

**Reply by The-Futurist** [reply]
Yes. It's documented as "Indicates the physical position of fields within the unmanaged representation of a class or structure". So my scenario doesn't involve marshaling, just manipulating unmanaged (not under GC control) memory using managed code. I think there's no support, just hardcoding 16 (in my case) which works fine. I just wanted the language (or runtime) to do more.

I want the option for the physical layout of a C# struct (naturally, only in narrow cases) to be defined by source code, it seems that this is something Microsoft are reticent to consider, a similar question also comes up with respect to bitfields. Where basically dealing with situations in which layout, alignment and padding ([even perhaps endianess](https://stackoverflow.com/a/55966787)) are a necessary part of a contract. 

I've recently done similar work with C# on an [STM32MP157 based board](https://github.com/Steadsoft/Endpoint/blob/main/Steadsoft.Spitfire.STM32MP153/Generated/TimerGeneral.cs), there I can replace bit fields with properties but the whole subject come up time and again when writing C# in these situations.

Control over such details is taken for granted in a systems programming language, but has been completely prevented in .Net.

**Reply by CyrusNajmabadi** [reply]
> I want the option for the physical layout of a C# struct (naturally, only in narrow cases) to be defined by source code

Again, why isn't that attribute sufficient?  It's for stating the physical layout of a C# struct...  could you clarify why that's not good enough here? Thanks!

**Reply by CyrusNajmabadi** [reply]
> Control over such details is taken for granted in a systems programming language, but has been completely prevented in .Net.

WDYM "has been completely prevented"?  Based on this:

> just manipulating unmanaged (not under GC control) memory using managed code. I think there's no support, just hardcoding 16 (in my case)

It sounds like you have an unmanaged type, and you have some value 16 bytes within it.  Why are the existing tools provided that allow specifying this sort of layout not sufficient for your use case?

> So my scenario doesn't involve marshaling,

I didn't think this was abotu marshaling.  But manipulating unmanaged memory is something .net does a fairly good job of, and has lots of tools and APIs to help out with.

> it seems that this is something Microsoft are reticent to consider,

I think the only reticence would be about adding something there is already support for.  If you can clairfy why the existing APIs are *insufficient* we can discuss what new APIs might be helpful.

**Reply by CyrusNajmabadi** [reply]
> a similar question also comes up with respect to bitfields. Where basically dealing with situations in which layout, alignment and padding

There's already a thread on that topic at Csharplang.  The view there was that there's no need for lang support as this is trivially supported by creation of a simple generator.  I believe a community member (or several) already ended up creating such a generator.  In the case of bitfields, it's a case where it's just very simple to accomplish, so we're not highly motivated to do expensive lang design work when there's a workable solution for the ecosystem that someone can pick up and use on any compiler/lang-version that supports generators.

Cheers :)

**Reply by CyrusNajmabadi** [reply]
> I just wanted the language (or runtime) to do more.

Can you be *very precise* as to the "do more" that you are referring to?  In particular, please indicate precisely what isn't possible right now, and which you then need dedicated APIs in order to support.  Based on the convos over at csharplang *and* here, it seems like the APIs are available, but they're not being used for some reason.  The "reason" part is what isn't clear currently.   Are they insufficient in some manner?  Is there something still missing you need?  Is it that they're not ergonomic enough?  More information is needed here because the parts you find absent aren't really being clearly stated. 

Thanks!

**Reply by The-Futurist** [reply]
The documentation refers to the "unmanaged _representation_ of a struct" I take that to mean something different to the managed representation of a struct, don't you? I interpret "representation" to encompass, order, size, offsets and padding if it means something else then kindly explain.

@tannergooding  regards these as bein different, read his comment here:

https://github.com/dotnet/runtime/discussions/100487#discussioncomment-8975642

@richlander please take note of the nature of this exchange between @CyrusNajmabadi and me, I really am getting fed up with his antics.

**Reply by CyrusNajmabadi** [reply]
Yes.  but as you mentioned: > So my scenario doesn't involve marshaling, just manipulating unmanaged (not under GC control) memory

:)

In this case, it seems like that is what you are asking for.  Right?

**Reply by The-Futurist** [reply]
Can you or anyone at Microsoft explain the difference between managed and unmanaged representations of a struct? do you claim there's no difference?

**Reply by CyrusNajmabadi** [reply]
Perhaps we can come at this from a different direction, since it does not seem as if my above questions were understood.  Can you give an example of: 

> So my scenario doesn't involve marshaling, just manipulating unmanaged (not under GC control) memory using managed code.

Please provide ane xample of such unmanaged memory.  Based on that, we can discuss FieldOffset and whether or not it would work for your case.  So far, afaict, it should work.  However, you seem to feel otherwise.  But lacking a real world case you have, it's hard for me to say one way or the other. :)

> I interpret "representation" to encompass, order, size, offsets and padding if it means something else then kindly explain.

Afaict, you were saying: > I think there's no support, just hardcoding 16 (in my case) which works fine. 

I don't know why yiou'd have to hardcode 16 (at least as far as your *code* goes).  To me, if you just specify the appropriate offsets in the `[FieldOffset(...)]` attribute, then things should work fine.  The runtime will respect that wrt to the location of the fields (that's literally *why* it is the "field offset attribute").  You can then use any of the field-offset calculation techniques listed in the threads already you've created on this topic.

--

Again, without an example, it's hard to state for certain.  But given your explanations so far, it seems like this would work.  I've tested this locally and it works fine for me.  So i'm curious if you've actually gone and tried this before dismissing it.  :)

**Reply by CyrusNajmabadi** [reply]
> I really am getting fed up with his antics.

@Korporal I don't know what to tell you.  I'm responding to your literal requests here, attempting to provide you with a solution to your problem.   I've provided code samples, ideas, suggestions, and links to available pieces of technology to try to solve things for you.  If you view that as 'antics', then that's on you.

**Reply by The-Futurist** [reply]
Please can someone answer my question, these are terms used in the documentation and I don't think there's even a consensus amongst Microsoft engineers - Can you or anyone at Microsoft explain the difference between managed and unmanaged representations of a struct? do you claim there's no difference?

**Reply by CyrusNajmabadi** [reply]
> do you claim there's no difference?

I'm claiming that, based on your description of the problem space, this attribute appears to be the right solution for your needs.  Have you tried it out?  Locally, i have, and everything seems to be working as i would expect, and my own local analog to yoru problem statement seems to work properly.

However, i'm not *100%* sure if that means it will work for you because i don't know your precise use case.  Do you have a code sample we can look at?  Do you have a definition for your structure?  If you can provide that, it would be very helpful.  I'm trying my best to provide a solution to your problem to the best of my ability, but i'm inherently limited as i only have some vague descriptions of what you're trying to do.

Finally, it's unclear to me if you've actually attempted any of the suggested approaches, or used any of the linked options, in order to determine if they work properly for your use case or not.  Based on your rejection of them, my guess is tha tyou have.  But it's really not clear.  

1. if you have tried these out, and they did not work, can you please give an example of something that didn't work properly?
2. if you have not tried them out... can you please actually do so?   Thanks :)

**Reply by The-Futurist** [reply]
Why won't you answer my question? it must have an answer, is there a difference yes or no?

**Reply by CyrusNajmabadi** [reply]
@Korporal I'm not sure why you're overindexing on these topics.  As you mentioned before, you're not doing marshalling ("So my scenario doesn't involve marshaling").  So you're not going to be using `[MarshalAs...]` or anything like that.  

So far, the only requirement i've seen relates to offsets, in which case `StructLayout.Explicit` and `FieldOffset` seem like the tools already available that solve that problem.  It's possible you have more requirements as well.  But i haven't heard them stated, so i'm just trying to provide you with the solution and options for that aspect of your problem space.

If you wanted to list out your requirements fully (optionally, and desirably, with some code examples showing what your structures look like), that would be helpful here.  But, for now, afaict, the links i've provided are the documented, well-supported, solutions to this particular problem.

Thanks! :)

**Reply by CyrusNajmabadi** [reply]
> Why won't you answer my question? 

As mentioned, i don't know why you're thinking that is relevant here :).  Are you marshaling your data with attributes like MarshalAs?  

To be able to answer, i need to know what i'm missing wrt to your problem domain.  A clear explanation of your requirements, along with some potential examples of your data structures would be super helpful.  Thanks very much!

if you can provide that, i should be able to answer clearly.  Cheers :)

>  is there a difference yes or no?

as  i *said*, based on the description *so far*, it doesn't seem relevant.  With more information, i can make a more informed answer here as *it depends*.  Now, i'm guessing that it won't matter.  But more info will help.  Tnx :)

**Reply by The-Futurist** [reply]
The documentation uses the phrase "unmanaged representation" I take that as implying that there's a difference between managed and unmanaged representations, so is there a difference? Let's have some input from the Runtime folks here, you suggested I ask them so that's what I'm doing.

**Reply by CyrusNajmabadi** [reply]
@Korporal  in the meantime, if you could provide answers to the questions asked about your requirements, and some sample data structure layouts, that would be terrific. It would make it possible to know that we're on the same page, and to ensure that offered solutions work.  Thanks very much! :)

Also, does this mean you haven't tried out these options?  It's still very unclear, so it's hard to respond wrt to any perceived shortcomings as i don't really know what issues you're running into (or if you're running into issues at all).  Thanks! :)

**Reply by tannergooding** [reply]
> @tannergooding regards these as bein different, read his comment here:

Yes, there are a few quirks around `managed` vs `unmanaged` layout and how various `StructLayout`/`FieldOffset` attributes work

There's a little bit of related discussion on that here and how the attributes were originally designed vs latter changes that came into place: https://github.com/dotnet/runtime/pull/100289#discussion_r1541422490

------------

There are, in general, a lot of quirks and problems that can occur when you start trying to manually manage your type layouts, particularly if you start trying to do things that would not be possible to define in C/C++. As an example, you can't define anything other than `FieldOffset(0)` in C/C++. The same goes for things like custom packing, which was touched on for C/C++ here: https://devblogs.microsoft.com/oldnewthing/20200103-00/?p=103290.

Type layout computation is dependent on the target ABI (which is defined by the target CPU and Operating System), it is not something that can be done by the C# compiler. The JIT/AOT compiler is able to do it, since it knows the target ABI being compiled against and for `blittable` types will directly match the corresponding layout from C/C++. For managed types, there are very few guarantees given and the JIT will generally ignore most layout attributes and simply treat it as `auto` instead. The main exception is `LayoutKind.Explicit` where it will match the specified offsets and throw a `TypeLoadException` if you've done something truly illegal (like overlapping a reference and value type). -- However, it's worth noting that just because a `TypeLoadException` isn't thrown, doesn't mean your code is well-defined. You can easily run into aliasing or other bugs if you're not treating it as a proper typed union where only 1 field is valid to access for a given piece of memory (which is also a requirement of unions in C/C++).

Because the JIT/AOT knows the type layout statically, due to it knowing the target ABI, things like `sizeof` are compiled down to constants. The JIT has some minimal pattern recognition for the general `offsetof` pattern as well and will often compile down to a constant or a near constant where possible. If there's patterns we don't recognize, that be an opportunity to file an issue asking for it to be optimized (the cost of caching for lookup, especially in something like a dictionary, will be more expensive than the 3-cycle `lea; lea; sub` that it would generate for dynamic computation in the worst case).

**Reply by tannergooding** [reply]
Typically speaking you should at worst be mirroring some type that was defined in C/C++ 20-30+ years ago and which can't be changed due to back-compat. Otherwise, you should be avoiding custom type layouts altogether. The worst case for C is then handling `unions` which never require more than `FieldOffset(0)` and anonymous structs/unions which are actually nested structs with. The simple example I give is generally the following, which shows how anonymous and inline fields differ:
```cpp
struct Derived1
{
    struct
    {
        uint64_t x;
        uint32_t y;
    } Base;

    uint32_t z;
};

struct Derived2
{
    uint64_t x;
    uint32_t y;

    uint32_t z;
};
```

Thus, the only guaranteed valid C# definition for `Derived1` is something like the following. Using `FieldOffset` or otherwise declaring inline fields can break on different platforms depending on how they've defined packing and other layout details to function. These types of breaks become particularly apparent for interop code which was written only for x86/x64 Windows when they start running on Arm32, Arm64, and Unix (Linux, MacOS, iOS, Android, etc) machines.
```csharp
public struct Base
{
    public ulong x;
    public uint y;
}

public struct Derived1
{
    public Base _Anonymous;
    public uint z;

    // You can optionally define helpers like
    [UnscopedRef]
    public ref ulong x => ref _Anonymous.x;

    [UnscopedRef]
    public ref uint y => ref _Anonymous.y;
}
```

**Reply by The-Futurist** [reply]
Thank you @tannergooding. The origin of this entire thread is, given a physical structure of memory layout, can we always define C# structs to correctly represent that layout. The physical layout of memory can be regarded as a contract, for example a COFF or ELF or GIF file contains numerous precisely ordered and laid out little structures. Does C# today allow me to always define a struct that can correctly describe and access such data? 

I want to stress that no interop is implied here, typically one would have an unsafe C# pointer of some struct type and expect to be able to use that to access the various byte fields without the .Net runtime imposing its own order, padding, offsets or even alignments.

Bit fields are out of scope too, I'm not concerned with these, only the freedom to define structs composed of blittable types and nest such structures. The documentation for `FieldOffsetAtribute` implies this only applies to "unmanaged representation" but I can find no precise definition for this term.

@jkotas - you might have views on this.

**Reply by tannergooding** [reply]
> Does C# today allow me to always define a struct that can correctly describe and access such data?

If you can define such a struct in C/C++, then you can define it in C# and it will work the same (provided you are only using blittable types). The C# struct definition will be essentially identical to the type define in C/C++, only using `struct { }` + `[FieldOffset(0)]` instead of `union { }`

If you can't define such a struct in C/C++, then you shouldn't be trying to define such a struct in C#. Instead, you should be using something more like a stream and binary reader/writer to ensure that the strict layout given is matched.

If you've ever looked at how something like the core PE reader used by the JIT or S.R.Metadata works, you'll note that it will use a struct in a few cases where that is possible. But there are many other cases it doesn't have a struct it can use and explicitly does binary reading to get the values it needs. There are also often concepts for these types of files like relative pointers which need helper APIs to compute where the next piece of data is read from.

> I want to stress that no interop is implied here

Interop is very much implied here as interop does not just pertain to `p/invoke`. Interop very much applies to any scenario where you're trying to ensure .NET code works correctly with something that is defined outside of .NET, including things like reading COFF, ELF, GIF, JPEG, or other file formats.

**Reply by CyrusNajmabadi** [reply]
> The physical layout of memory can be regarded as a contract, for example a COFF 

For COFF reading, could you use [PEHeaders](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.portableexecutable.peheaders?view=net-8.0)?

**Reply by CyrusNajmabadi** [reply]
@tannergooding Note: i think @Korporal has a fixed target platform he intends to run on.  So he may be willing to take a solution that is not fully portable, but works there alone.  Though it would be good to get clarification from him on this.

**Reply by jkotas** [reply]
> @jkotas - you might have views on this.

I do not have anything to add to @tannergooding's answers. Tanner covered every aspect of this problem space that I can think of.
